packages/opentelemetry-sdk-trace-base/src/export/BatchSpanProcessorBase.ts

```diff
@@ -179,20 +172,43 @@ export abstract class BatchSpanProcessorBase<T extends BufferConfig>
     });
   }
 
-  private _maybeStartTimer() {
-    if (this._timer !== undefined) return;
-    this._timer = setTimeout(() => {
-      this._flushOneBatch()
-        .then(() => {
-          if (this._finishedSpans.length > 0) {
-            this._clearTimer();
-            this._maybeStartTimer();
-          }
-        })
-        .catch(e => {
-          globalErrorHandler(e);
-        });
-    }, this._scheduledDelayMillis);
+  private async _tryExportOneBatch(): Promise<void> {
+    try {
+      await this._exportOneBatch();
+    } catch (e) {
+      globalErrorHandler(e);
+    }
+  }
+
+  private async _export() {
+    await this._tryExportOneBatch();
+
+    if (this._finishedSpans.length >= this._maxExportBatchSize) {
+      this._resetTimer(0);
+    } else {
+      this._resetTimer(this._scheduledDelayMillis);
+    }
+  }
+
+  private _exportCompleteBatch() {
+    if (this._finishedSpans.length < this._maxExportBatchSize) {
+      return;
+    }
+
+    if (this._nextExport === 0) {
```

Comment:
Personally, I'm not a fan of this pattern of "detecting" whether we should start a timer or not -- I perfer more explicit start/reset descriptions.

It took me a while to understand that this it the "flag" you are effectively using to determine whether there is "more" data (or not) and then whether to "reset" the timer.

I also don't like that you can "reset" an already running timer, as it could be very easy for someone to come along later and cause an endless (until it got full) delay if something is getting batched at a regular interval.

Reply:
`_nextExport` is only used to avoid needlessly resetting the timer, i.e. it means that an export is already queued next cycle. Think about appending thousands of spans consecutively in the same event loop cycle.

What would your alternative be to resetting an already running timer? Starting an export in a new promise right away when the buffer reaches the threshold can't be done as it would cause both too many concurrent exports and would nullify the concept of max queue size.

Reply:
> _nextExport is only used to avoid needlessly resetting the timer, i.e. it means that an export is already queued next cycle.

this._timer will (should) be undefined when no timer is already running :smile: (as long as you also set the value to undefined (or null) within the setTimeout() implementation as well.

Reply:
Which is basically what the previous implementation was doing with `if (this._timer !== undefined) return;` in the _maybeSet function (although I'd also prefer to see a check that nothing is currently batched as well -- to avoid creating the timer in the first place.

Reply:
The timer is now always running, basically this new implementation sets the timeout to 0 once the batch size is exceeded.

Reply:
But if there is nothing in the batch we should not have any timer running...
ie. The timer should only be running if there is something waiting to be sent, otherwise, if an application is sitting doing nothing (because the user walked away) by having a running timer this can cause the device (client) to never go to sleep and therefore use more resource (power / battery) when not necessary

Reply:
Given the spec (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#batching-processor) we could probably simplify this whole processor to just be a `setInterval` and a list of spans. During the most recent spec iteration it was made clear that there is no need to wait for the previous export to complete before starting another one. The word "returns" is used very explicitly in the spec and refers to thread safety, not to the actual async task of the export.

Reply:
Wouldn't it invalidate the `maxQueueSize` parameter in this case? E.g. when starting an export when the batch size has been reached